### PR TITLE
input: add a gpio based keyboard matrix driver

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Initial gpio based driver for the now generalized keyboard scanning library. Should be useful to tweak and troubleshoot it the generic keyboard scanning code on any random devboard. Just tested it on a random board shorting pins out, got some dedicated hardware to do more testing on the way. Prob deserves some better documentation but I have more features and options for this in the pipeline so I think I'd work on a nice doc page at a later stage.

---

Add a GPIO based keyboard matrix driver using the generic keyboard matrix code.

### CODE REVIEW

File: tests/drivers/build_all/input/app.overlay [hunk 1]
```diff
@@ -26,6 +26,17 @@
...
 
+		kbd-matrix {
+			compatible = "gpio-kbd-matrix";
+			row-gpios = <&gpio0 0 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>,
+				    <&gpio0 0 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
+			col-gpios = <&gpio0 0 GPIO_ACTIVE_LOW>,
+				    <&gpio0 0 GPIO_ACTIVE_LOW>,
+				    <&gpio0 0 GPIO_ACTIVE_LOW>;
+			row-size = <2>;
+			col-size = <3>;
```
Comment (gmarull) [inline tests/drivers/build_all/input/app.overlay:37]: couldn't these be inferred from len(row-gpios) and len(col-gpios)?

Comment (fabiobaltieri) [reply tests/drivers/build_all/input/app.overlay:37]: not really, they are defined in the generic code while the gpios one are in this specific implementation

Comment (gmarull) [reply tests/drivers/build_all/input/app.overlay:37]: I mean, the driver could make these optional in this particular case?

Comment (fabiobaltieri) [reply tests/drivers/build_all/input/app.overlay:37]: ...unless I duplicate the initializer and make a version that takes row and col as an argument, how does that sound?

Comment (gmarull) [reply tests/drivers/build_all/input/app.overlay:37]: up to you, just thought these are a bit redundant in this case.

File: drivers/input/input_gpio_kbd_matrix.c [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	};											\
+	static const struct gpio_dt_spec gpio_kbd_matrix_col_gpio_##n[DT_INST_PROP_LEN(		\
+			n, col_gpios)] = {							\
+		DT_INST_FOREACH_PROP_ELEM_SEP(n, col_gpios, GPIO_DT_SPEC_GET_BY_IDX, (,))	\
+	};											\
+	static struct gpio_callback gpio_kbd_matrix_gpio_cb_##n[DT_INST_PROP_LEN(n, row_gpios)];\
+												\
+	static const struct gpio_kbd_matrix_config gpio_kbd_matrix_cfg_##n = {			\
+		.common = INPUT_KBD_MATRIX_DT_INST_COMMON_CONFIG_INIT_ROW_COL(			\
+			0, gpio_kbd_matrix_api,							\
```
Comment (fabiobaltieri) [inline drivers/input/input_gpio_kbd_matrix.c:184]: sneaky `0`

File: tests/drivers/build_all/input/app.overlay (hunk 1 above)
Comment (fabiobaltieri) [reply tests/drivers/build_all/input/app.overlay]: alright how's it look now? more macros more better

File: dts/bindings/input/gpio-kbd-matrix.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                      <&gpio0 3 GPIO_ACTIVE_LOW>,
+                      <&gpio0 4 GPIO_ACTIVE_LOW>;
+          row-size = <2>;
+          col-size = <3>;
+          no-ghostkey-check;
+  };
+
+compatible: "gpio-kbd-matrix"
+
+include: kbd-matrix-common.yaml
```
Comment (gmarull) [inline dts/bindings/input/gpio-kbd-matrix.yaml:25]:
```suggestion
          no-ghostkey-check;
  };

compatible: "gpio-kbd-matrix"

include: kbd-matrix-common.yaml
```

maybe filter-out row-size/col-size when including?

Comment (fabiobaltieri) [reply dts/bindings/input/gpio-kbd-matrix.yaml]: done, good idea, also done

**Review by gmarull — DISMISSED** [review]

File: drivers/input/input_gpio_kbd_matrix.c [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	struct input_kbd_matrix_common_config common;
+	const struct gpio_dt_spec *row_gpio;
+	const struct gpio_dt_spec *col_gpio;
+	struct gpio_callback *gpio_cb;
+	gpio_callback_handler_t handler;
+};
+
+struct gpio_kbd_matrix_data {
+	struct input_kbd_matrix_common_data common;
+	int last_col_state;
```
Comment (fabiobaltieri) [inline drivers/input/input_gpio_kbd_matrix.c:32]: this should really be a uint32

**Comment by fabiobaltieri** [comment]
@gmarull sneaked in one more tweak post approval, the col limit of 32 is beacuse of the last state in data, made in an `uint32_t` explicitly.

**Review by gmarull — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
Turns out I had messed up the common kbd config struct and had the api struct itself rather than a point to it, clang-16 fired up a (fairly cryptic) warning about it, guess it was working by copying the structure fields at init time (didn't even know it was possible). Slipped in another commit to fix that as well.

**Review by gmarull — DISMISSED** [review]

File: tests/drivers/build_all/input/app.overlay [hunk 5]
```diff
@@ -26,6 +26,15 @@
 			};
 		};
 
+		kbd-matrix {
+			compatible = "gpio-kbd-matrix";
+			row-gpios = <&gpio0 0 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>,
+				    <&gpio0 0 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
+			col-gpios = <&gpio0 0 GPIO_ACTIVE_LOW>,
+				    <&gpio0 0 GPIO_ACTIVE_LOW>,
+				    <&gpio0 0 GPIO_ACTIVE_LOW>;
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply tests/drivers/build_all/input/app.overlay]: Ok, I  changed it to sequential numbers. Spotted some other incoherences in this file btw, I'll fix the rest separately.

**Comment by fabiobaltieri** [comment]
Ah, doxygen is missing the file, guess it should have been in `include/` anyway, moved, let's see if it like it better.

File: dts/bindings/input/gpio-kbd-matrix.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      - row-size
+      - col-size
+
+properties:
+  row-gpios:
+    type: phandle-array
+    required: true
+    description: |
+      GPIO for the keyboard matrix rows, must match the count in row-size,
+      supports up to 8 rows, all row GPIO pins must have interrupt support.
```
Comment (keith-zephyr) [inline dts/bindings/input/gpio-kbd-matrix.yaml:35]: Drop the "match row-size" language.
```suggestion
      GPIO for the keyboard matrix rows. All row GPIO pins must have interrupt support.
```

File: dts/bindings/input/gpio-kbd-matrix.yaml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+  col-gpios:
+    type: phandle-array
+    required: true
+    description: |
+      GPIO for the keyboard matrix columns, must match the count in col-size,
+      supports up to 32 columns.
```
Comment (keith-zephyr) [inline dts/bindings/input/gpio-kbd-matrix.yaml:42]: Drop "match col-size" languate.

Comment (fabiobaltieri) [reply dts/bindings/input/gpio-kbd-matrix.yaml]: Thanks! Done, added a note on column gpio driving as well.

File: dts/bindings/input/gpio-kbd-matrix.yaml (hunk 6 above)
Comment (fabiobaltieri) [reply dts/bindings/input/gpio-kbd-matrix.yaml]: Done, I kept the note about the limit of 8 as that still applies, it's enforced in the common code.

**Review by keith-zephyr — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
Ok I bumped into few more doxygen issues after checking https://builds.zephyrproject.io/zephyr/pr/65117/docs/services/input/index.html#keyboard-matrix-api-reference, it should be ok now.

Could probably do a better job at rewriting the commit history but I'd rather spare myself the rebase conflicts, this PR took few unexpected turns, can't complain about the end result though, many things needed fixing.

Thanks for your patience on the reviews.

**Review by keith-zephyr — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Pull request merged**